### PR TITLE
Support set_model for callable functions in pyfunc

### PR DESCRIPTION
**Pull request by annzhang-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/annzhang-db/mlflow/pull/12035?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12035/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12035
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

- Allow for callable functions to go through set_model and load properly from a file

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/pyfunc/test_model_export_with_class_and_artifacts.py [hunk 1]
```diff
@@ -1683,7 +1683,7 @@ def test_model_log_with_resources(tmp_path):
 def test_pyfunc_as_code_log_and_load():
     with mlflow.start_run():
         model_info = mlflow.pyfunc.log_model(
-            python_model="tests/pyfunc/sample_code/code.py",
```
Comment (annzhang-db) [inline tests/pyfunc/test_model_export_with_class_and_artifacts.py:1686]: Renamed python files for improved clarity.

File: mlflow/pyfunc/model.py [hunk 2]
```diff
@@ -513,7 +519,6 @@ def _load_context_model_and_signature(
 
     context = PythonModelContext(artifacts=artifacts, model_config=model_config)
     python_model.load_context(context=context)
-    signature = mlflow.models.Model.load(model_path).signature
```
Comment (annzhang-db) [inline mlflow/pyfunc/model.py:516]: Moved up so we can use this for callable functions.

File: mlflow/pyfunc/model.py [hunk 3]
```diff
@@ -473,11 +473,17 @@ def _load_context_model_and_signature(
...
 
     if MODEL_CODE_PATH in pyfunc_config:
         conf_model_code_path = pyfunc_config.get(MODEL_CODE_PATH)
         model_code_path = os.path.join(model_path, os.path.basename(conf_model_code_path))
         python_model = _load_model_code_path(model_code_path, model_config)
+
+        if callable(python_model):
+            # hints initialized to None, but _get_type_hints will extract them from the function
+            hints = None
+            python_model = _FunctionPythonModel(python_model, hints, signature)
```
Comment (harupy) [inline mlflow/pyfunc/model.py:486]:
```suggestion
            python_model = _FunctionPythonModel(python_model, signature=signature)
```

**Review by harupy — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/12035/files#r1606150349 is addressed!

File: tests/pyfunc/sample_code/predict_code.py
Comment (harupy) [inline tests/pyfunc/sample_code/predict_code.py:1]: can we rename this file  `functional_model.py`?

File: mlflow/pyfunc/model.py (hunk 3 above)
Comment (harupy) [reply mlflow/pyfunc/model.py:486]: It looks the `hints` attribute is not used.

Comment (harupy) [reply mlflow/pyfunc/model.py:486]: Not a blocker, can be fixed in a follow-up.

**Pull request merged**